src/sources/aws_s3/sqs.rs

```diff
@@ -0,0 +1,517 @@
+                            line_agg::Logic::new(config.clone()),
+                        )
+                        .map(|(_src, line, _context)| line),
+                    );
+                }
+
+                let stream = lines.filter_map(|line| {
+                    let mut event = Event::from(line);
+
+                    let log = event.as_mut_log();
+                    log.insert("bucket", s3_event.s3.bucket.name.clone());
+                    log.insert("object", s3_event.s3.object.key.clone());
+                    log.insert("region", s3_event.aws_region.clone());
+                    log.insert(log_schema().timestamp_key(), timestamp);
+
+                    if let Some(metadata) = &metadata {
+                        for (key, value) in metadata {
+                            log.insert(key, value.clone());
+                        }
+                    }
+
+                    futures::future::ready(Some(Ok(event)))
+                });
+
+                out.send_all(Compat::new(Box::pin(stream)))
+                    .compat()
+                    .await
+                    .map_err(|error| {
+                        error!(message = "Error sending S3 Logs", %error);
+                    })
+                    .ok();
+
+                read_error
+                    .map(|error| {
+                        Err(ProcessingError::ReadObject {
+                            error,
+                            bucket: s3_event.s3.bucket.name.clone(),
+                            key: s3_event.s3.object.key.clone(),
+                        })
+                    })
+                    .unwrap_or(Ok(()))
+            }
+            None => Ok(()),
+        }
+    }
+
+    async fn receive_messages(&self) -> Result<Vec<Message>, RusotoError<ReceiveMessageError>> {
+        self.sqs_client
+            .receive_message(ReceiveMessageRequest {
+                queue_url: self.queue_url.clone(),
+                max_number_of_messages: Some(10),
+                visibility_timeout: Some(self.visibility_timeout_secs),
+                ..Default::default()
+            })
+            .map_ok(|res| res.messages.unwrap_or_default())
+            .await
+    }
+
+    async fn delete_message(
+        &self,
+        receipt_handle: String,
+    ) -> Result<(), RusotoError<DeleteMessageError>> {
+        self.sqs_client
+            .delete_message(DeleteMessageRequest {
+                queue_url: self.queue_url.clone(),
+                receipt_handle,
+            })
+            .await
+    }
+}
+
+// https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "PascalCase")]
+struct S3Event {
+    records: Vec<S3EventRecord>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct S3EventRecord {
+    event_version: String, // TODO compare >=
```

Comment:
Is this still TODO?

Reply:
Ah, yes, thanks. I missed that when combing for lingering TODOs.